### ISSUE TITLE
Parallel assertions, cocotb terminates incorrectly 

### ISSUE THREAD

**Issue by aqeelmahroof** [cocotb/cocotb#3062]
Hi , When assertions (or exceptions) raised in parallel co-routines at the same time , cocotb is not terminating correctly 

Here two co-routines  are started and raising exception after 20 units of time 
```
async def call_error(waiter):
    await waiter
    assert False

@cocotb.test(skip=False)
async def cocotb_check(dut):
    waiter = Timer(20)
    for i in range(2):
        cocotb.start_soon(call_error(waiter))
    await Timer(30)    
```
In simulation the cocotb raises InternalError and terminates
This test-bench has 5 test cases, and the 3rd one has the issue, the cocotb terminates at 3rd test and not executing the other tests  
```
# Loading C:/Apps/WPy64-3940/python-3.9.4.amd64/lib/site-packages/cocotb/libs/cocotbfli_modelsim.dll
#      -.--ns INFO     cocotb.gpi                         ..ib\fli\FliImpl.cpp:1105 in cocotb_init                     cocotb_init called
#      -.--ns INFO     cocotb.gpi                         ..mbed\gpi_embed.cpp:78   in set_program_name_in_venv        Did not detect Python virtual environment. Using system-wide Python interpreter
#      -.--ns INFO     cocotb.gpi                         ..\gpi\GpiCommon.cpp:99   in gpi_print_registered_impl       FLI registered
#      0.00ns INFO     Running on ModelSim for QuestaIntel Starter FPGA Edition-64 version 2022.1 2022.01
#
#      0.00ns INFO     Running tests with cocotb v1.6.2 from C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb
#
#      0.00ns INFO     Seeding Python random module with 1661840909
#
#      0.00ns INFO     Found test cocotb_check.cocotb_pass
#
#      0.00ns INFO     Found test cocotb_check.cocotb_check
#
#      0.00ns INFO     Found test cocotb_check.cocotb_check2
#
#      0.00ns INFO     Found test cocotb_check.cocotb_check3
#
#      0.00ns INFO     Found test cocotb_check.cocotb_check4
#
#      0.00ns INFO     running cocotb_pass (1/5)
#
ERROR: called callback function threw exception
Traceback (most recent call last):
  File "\check_cocotb_error\cocotb_check.py", line 7, in call_error
    assert False
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\scheduler.py", line 357, in _react
    self._event_loop(trigger)
  File "C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\scheduler.py", line 451, in _event_loop
    self._schedule(coro, trigger=trigger)
  File "C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\scheduler.py", line 897, in _schedule
    self._unschedule(coroutine)
  File "C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\scheduler.py", line 537, in _unschedule
    self._abort_test(e)
  File "C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\scheduler.py", line 955, in _abort_test
    raise InternalError("Outcome already has a value, but is being set again.")
cocotb.scheduler.InternalError: Outcome already has a value, but is being set again.
#     31.00ns INFO     cocotb_pass passed
#
#     31.00ns INFO     running cocotb_check (2/5)
#
#     51.00ns INFO     Test stopped by this forked coroutine
#
#     52.00ns INFO     cocotb_check failed
#
#                      Traceback (most recent call last):
#
#                        File "\check_cocotb_error\cocotb_check.py", line 7, in call_error
#
#                          assert False
#
#                      AssertionError
#
#     52.00ns INFO     running cocotb_check2 (3/5)
#
#     72.00ns INFO     Test stopped by this forked coroutine
#
#     72.00ns INFO     Test stopped by this forked coroutine
#
# Simulation halt requested by foreign interface.
# End time: 16:28:31 on Aug 30,2022, Elapsed time: 0:00:07
# Errors: 0, Warnings: 1
ERROR: results.xml was not written by the simulation!
make[1]: *** [/c/Apps/WPy64-3940/python-3.9.4.amd64/lib/site-packages/cocotb/share/makefiles/simulators/Makefile.questa:159: results.xml] Error 1
make[1]: Leaving directory '/check_cocotb_error'
make: *** [C:\Apps\WPy64-3940\python-3.9.4.amd64\lib\site-packages\cocotb\share\makefiles/Makefile.inc:40: sim] Error 2
```
Futher when i comment this line thing works as expected 
https://github.com/cocotb/cocotb/blob/997ee6dd906ea19954cb37f75a1406874b518db7/cocotb/scheduler.py#L1024 

Is this the expected behaviour ?

testbench: [check_cocotb_error.zip](https://github.com/cocotb/cocotb/files/9450418/check_cocotb_error.zip)
logfile:  [sim.log](https://github.com/cocotb/cocotb/files/9450470/sim.log)
vhdl
cocotb 1.6.2
windows 64bit 
Questa 2022.1
python 3.9.4

**Comment by marlonjames** [comment]
This is a duplicate of #1347.
